samples/hello_fpga.c

```diff
@@ -230,6 +231,33 @@ fpga_result get_bus(fpga_token tok, uint8_t *bus)
 	return res1 != FPGA_OK ? res1 : res2;
 }
 
+/* Is the FPGA simulated with ASE? */
+bool probe_for_ase()
```

Comment:
Could you also add a unit test for this fn to testing/hello_fpga/test_hello_fpga_c.cpp?

Reply:
No. I'm tired of being the one fixing the DCP regression suite. Please work with them and get someone to maintain these.

Reply:
This isn't the DCP regression suite.  The tests I'm referring to are the unit tests for OPAE.  Your submission causes the coverage to drop (Coverage decreased (-0.05%) to 78.722%).  We try to at least maintain the current coverage with new submissions.  Sorry, but your code will not be accepted without abiding by the submission guidelines.

Reply:
I read the unit test code and have no idea how a unit test for this would be structured. The code always runs, so it is executed by mock_hello_fpga_c_p.

I don't know how the automated process could have made any conclusions about coverage decreasing. There is a conditional block in probe_for_ase(), but there's no way to write a unit test that forces it down a path in the mock environment. The same is true for the new loop at the end waiting for traffic to stop.

The actual behavior, which is more important than trying to convince the mock environment to do something, is quite well tested within the HW simulation environment and within the DCP ASE testing against NLB0 RTL.

Reply:
I think all the needs to be done is to add the function signature for probe_for_ase() to the top of the test file inside the extern "C" { block; then add a test case which calls the function.  That should be sufficient to keep the coverage unchanged.

Reply:
Huh? probe_for_ase() is called unconditionally by main() and main() is already probed by the tests.

Reply:
I see what you mean.  I suppose it would require too much mocking to get things to flow into the branch on 249.  OK with this one going in.

Reply:
Right. The unit tests are fine for testing argument parsing, storage leaks, etc. Not so good when it finally comes time to interact with the HW.

I bet that some semantically unreachable parts of the OPAE SDK could be better tested within the HW infrastructure tests. Something to consider for the future.

That first bug (the failure to destroy a token before overwriting it) is, theoretically, discoverable automatically. Have you thought about that direction instead of by inspection?

Reply:
Actually, yes.  We do have a valgrind component of our testing, though it's not automated and is a bit of a pain to deal with.  Automation would certainly be better.